### ISSUE TITLE
Wrapping a C++ program that uses the Adept autodiff library

### ISSUE THREAD

**Issue by Huzaifg** [swig/swig#2601]
I am struggling to wrap a C++ code that uses the [Adept automatic differentiation library](https://github.com/rjhogan/Adept-2). Adept works on operator overloading by instantiating a function template with a custom type which stores not only a variable but also a value of its derivative. In Adept, this custom type is `adouble`. 

I have a Vehicle Model that is defined [here](https://github.com/Huzaifg/temp/blob/main/d4dof/fourDof.cpp) and declared in [this header file](https://github.com/Huzaifg/temp/blob/main/d4dof/fourDof.h). This model, standalone can be built using this [MakeFile](https://github.com/Huzaifg/temp/blob/main/d4dof/Makefile). It is templatized to work with both `adouble`s and `doubles`. I want to wrap these functions, which internally use Adept to compute gradients and store them in the right places without having to wrap the entire Adept Library.

I create a interface file, as described in the documentation - this can be seen [here](https://github.com/Huzaifg/temp/blob/main/d4dof/interfaces/rom4.i) .  Since each of the functions take a `adept::Stack` object, I wrote [a wrapper class ](https://github.com/Huzaifg/temp/blob/main/d4dof/StackWrapper.h) so that I can create this stack object behind the scenes when I use Python. also have a [CMake file](https://github.com/Huzaifg/temp/blob/main/d4dof/interfaces/CMakeLists.txt) to build and link the Python Libs and the Adept libs to generate the Python module. I however get the following error - 

```
[ 25%] Swig compile rom4.i for python
[ 25%] Built target rom4_swig_compilation
[ 50%] Building CXX object CMakeFiles/_rom4.dir/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx.o
In file included from /usr/local/include/adept.h:17,
                 from /home/unjhawala/projectlets/model-repo/simple-vehicles/lang-c/dROMs/playground/d4dof/interfaces/../fourDof.h:6,
                 from /home/unjhawala/projectlets/model-repo/simple-vehicles/lang-c/dROMs/playground/d4dof/interfaces/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx:3667:
/usr/local/include/adept/Active.h: In instantiation of ‘adept::Active<T>::Active(const PType&, typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type*) [with PType = swig::SwigPySequence_Ref<adept::Active<double> >; Type = double; typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type = void]’:
/home/unjhawala/projectlets/model-repo/simple-vehicles/lang-c/dROMs/playground/d4dof/interfaces/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx:4940:30:   required from ‘void swig::assign(const SwigPySeq&, Seq*) [with SwigPySeq = SwigPySequence_Cont<adept::Active<double> >; Seq = std::vector<adept::Active<double> >]’
/home/unjhawala/projectlets/model-repo/simple-vehicles/lang-c/dROMs/playground/d4dof/interfaces/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx:4962:12:   required from ‘static int swig::traits_asptr_stdseq<Seq, T>::asptr(PyObject*, sequence**) [with Seq = std::vector<adept::Active<double> >; T = adept::Active<double>; PyObject = _object; sequence = std::vector<adept::Active<double> >]’
/home/unjhawala/projectlets/model-repo/simple-vehicles/lang-c/dROMs/playground/d4dof/interfaces/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx:5023:52:   required from ‘static int swig::traits_asptr<std::vector<T> >::asptr(PyObject*, std::vector<T>**) [with T = adept::Active<double>; PyObject = _object]’
/home/unjhawala/projectlets/model-repo/simple-vehicles/lang-c/dROMs/playground/d4dof/interfaces/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx:3987:37:   required from ‘int swig::asptr(PyObject*, Type**) [with Type = std::vector<adept::Active<double> >; PyObject = _object]’
/home/unjhawala/projectlets/model-repo/simple-vehicles/lang-c/dROMs/playground/d4dof/interfaces/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx:7549:23:   required from here
/usr/local/include/adept/Active.h:86:9: error: cannot convert ‘const swig::SwigPySequence_Ref<adept::Active<double> >’ to ‘double’ in initialization
   86 |       : val_(rhs), gradient_index_(ADEPT_ACTIVE_STACK->register_gradient())
      |         ^~~~~~~~~
make[2]: *** [CMakeFiles/_rom4.dir/build.make:76: CMakeFiles/_rom4.dir/CMakeFiles/_rom4.dir/rom4PYTHON_wrap.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/_rom4.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

Can someone help me understand the error and suggest ways to circumvent it? I think its associated with the `adouble` type that swig is unable to wrap. Is there some other way that I should be doing this?

I am running
OS: Arch Linux x86_64
Kernel: 6.2.13-arch1-1
g++ (GCC) 13.1.1

**Comment by wsfulton** [comment]
I suggest you post a very minimal example of this problem for SWIG users to debug. Some advise which might help: make sure adouble is fully defined for SWIG and avoid using typedefs in %template.

**Comment by Huzaifg** [comment]
Sorry for the elaborate example. Here is a simple one
Consider the following functions defined in the header _**solveODE.h**_

```
#ifndef SOLVEODE_H
#define SOLVEODE_H

#include "adept.h"

// Some functions that need to be wrapped
void computeStateJac(std::vector<adept::adouble>& v_st, std::vector<adept::adouble>& old_state, adept::Stack& stack);
void integrate(std::vector<adept::adouble>& v_st, std::vector<adept::adouble>& v_param, std::vector<adept::adouble>& controls, adept::Stack& stack);
void solve(std::vector<adept::adouble>& v_st, std::vector<adept::adouble>& v_params, std::vector<adept::adouble>& controls, std::vector<double> statejac, adept::Stack& stack);


#endif
```
These functions are defined in _**solveODE.cpp**_
```
#include "solveODE.h"
#include <cmath>
#include <iostream>
#include <fstream>
#include <algorithm>



using namespace adept;


void computeStateJac(std::vector<adouble>& v_st, std::vector<adouble>& old_state, std::vector<double> stateJac, adept::Stack& stack){
    stack.dependent(v_st[0]);
    stack.dependent(v_st[1]);
    stack.dependent(v_st[2]);
    stack.dependent(v_st[3]);
    stack.independent(&old_state[0], 4);
    stack.jacobian(&stateJac[0]); // Compute and store the State jacobian from the stack
    
    // Once computed, clear dependent and independent as we might need to compute some other jacobian
    stack.clear_dependents();
    stack.clear_independents();

}


// given the olds states, parmaeters and controls, integrate to the next time step using explicit Euler
void integrate(std::vector<adept::adouble>& v_st, std::vector<adept::adouble>& old, std::vector<adept::adouble>& v_param, std::vector<adept::adouble>& controls, adept::Stack& stack){
    v_st[0] = old[0] + cos(old[2])*old[3]*v_param[8];
    v_st[1] = old[1] + sin(old[2])*old[3]*v_param[8];
    v_st[2] = old[2] + (tan(controls[1]) / v_param[0])*old[3]*v_param[8]; // theta update

    adouble rGamma = v_param[3] * v_param[5];
    adouble f1 = v_param[6] * controls[2] - (v_param[6] * old[3] / (v_param[7] * rGamma));
    adouble f0 = old[3] * v_param[1] / rGamma + v_param[2];
    v_st[3] = old[3] +  ((rGamma / v_param[4]) * (f1 - f0))*v_param[8]; // V update
}

// Solve forward and store the state jacobian
void solve(std::vector<adept::adouble>& v_st, std::vector<adept::adouble>& v_params, std::vector<adept::adouble>& controls, std::vector<double> stateJac, adept::Stack& stack){

    std::vector<adouble> old(4);
    old[0] = v_st[0];
    old[1] = v_st[1];
    old[2] = v_st[2];
    old[3] = v_st[3];

    stack.new_recording();

    integrate(v_st, old, v_params, controls, stack);
    computeStateJac(v_st, old, stateJac, stack); // Compute and store the Jacobin
}
```
These functions are the ones I want to wrap. They however take some uncommon arguments such as a vector of `adouble`'s and a `adept::Stack` class. Both of these are from an external Automatic Differentiation library Adept. `adouble`'s is nothing but a typedef on `Active<double>` which is defined [here](https://github.com/rjhogan/Adept-2/blob/master/include/adept/Active.h).  
I tried wrapping these functions using this interface file _**simpleExample.i**_

```
%module simpleExample

%include "std_string.i"
%include "std_vector.i"

%{
#include "adept.h"
#include "solveODE.h"
%}
%include "solveODE.h";

%template(vector_adouble) std::vector <adept::adouble>;
%template(vector_double) std::vector <double>;
```

I however get the same error as above.
I think this is related to the fact that `adouble` is not fully defined for SWIG as @wsfulton said. But I do not know how to fully define it without having to wrap the entire library.  I just started out with SWIG and am probably missing something obvious. 

Any help is appreciated!

**Comment by ojwb** [comment]
Try adding `%import "adept.h"` to your interface file.

**Comment by Huzaifg** [comment]
I tried this by modifying my interface file to 

```
%module simpleExample

%include "std_string.i"
%include "std_vector.i"

%{
#include "adept.h"
#include "solveODE.h"
%}
%import "adept.h";
%include "solveODE.h";

%template(vector_adouble) std::vector <adept::adouble>;
%template(vector_double) std::vector <double>;
```

I still get the same error

```
[ 25%] Swig compile simpleExample.i for python
[ 25%] Built target simpleExample_swig_compilation
[ 50%] Building CXX object CMakeFiles/_simpleExample.dir/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx.o
In file included from /usr/local/include/adept.h:17,
                 from /home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:3655:
/usr/local/include/adept/Active.h: In instantiation of ‘adept::Active<T>::Active(const PType&, typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type*) [with PType = swig::SwigPySequence_Ref<adept::Active<double> >; Type = double; typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type = void]’:
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:4760:30:   required from ‘void swig::assign(const SwigPySeq&, Seq*) [with SwigPySeq = SwigPySequence_Cont<adept::Active<double> >; Seq = std::vector<adept::Active<double> >]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:4782:12:   required from ‘static int swig::traits_asptr_stdseq<Seq, T>::asptr(PyObject*, sequence**) [with Seq = std::vector<adept::Active<double> >; T = adept::Active<double>; PyObject = _object; sequence = std::vector<adept::Active<double> >]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:4843:52:   required from ‘static int swig::traits_asptr<std::vector<T> >::asptr(PyObject*, std::vector<T>**) [with T = adept::Active<double>; PyObject = _object]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:3807:37:   required from ‘int swig::asptr(PyObject*, Type**) [with Type = std::vector<adept::Active<double> >; PyObject = _object]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:6346:23:   required from here
/usr/local/include/adept/Active.h:86:9: error: cannot convert ‘const swig::SwigPySequence_Ref<adept::Active<double> >’ to ‘double’ in initialization
   86 |       : val_(rhs), gradient_index_(ADEPT_ACTIVE_STACK->register_gradient())
      |         ^~~~~~~~~
make[2]: *** [CMakeFiles/_simpleExample.dir/build.make:76: CMakeFiles/_simpleExample.dir/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/_simpleExample.dir/all] Error 2
make: *** [Makefile:91: all] Error 2

```
Then I also tried importing the `Active.h` file that contains the definations of the new types by doing this

```
%module simpleExample

%include "std_string.i"
%include "std_vector.i"

%{
#include "adept.h"
#include "solveODE.h"
%}
%import "adept.h";
%import <adept/Active.h>
%include "solveODE.h";

%template(vector_adouble) std::vector <adept::adouble>;
%template(vector_double) std::vector <double>;
```
I then however get this as the error

```
/usr/local/include/adept/Active.h:203: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:227: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:250: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:255: Error: Syntax error in input(3).
make[2]: *** [CMakeFiles/simpleExample_swig_compilation.dir/build.make:76: CMakeFiles/_simpleExample.dir/simpleExamplePYTHON.stamp] Error 1
make[2]: *** Deleting file 'CMakeFiles/_simpleExample.dir/simpleExamplePYTHON.stamp'
make[1]: *** [CMakeFiles/Makefile2:111: CMakeFiles/simpleExample_swig_compilation.dir/all] Error 2

```

**Comment by ojwb** [comment]
Yes, you'll need to import dependent headers too (or use `-importall` but that will probably only work if your headers don't include **any** system headers as otherwise it tries to follow `#include` into the system headers and that rarely works out well).

Line 255 is:

```
typename internal::enable_if<E::is_active && E::rank==0, Active&>::type
```

I think you're hitting #2228 there.

**Comment by Huzaifg** [comment]
Thanks, this does help. I modified by interface file to manually import all (I think) the dependent headers:
```
%module simpleExample

%include "std_string.i"
%include "std_vector.i"

%{
#include "adept.h"
#include "solveODE.h"
%}
%import "adept.h";
%import <adept/Active.h>
%import <adept/settings.h>
%import <adept/UnaryOperation.h>
%import <adept/BinaryOperation.h>
%import <adept/scalar_shortcuts.h>
%include "solveODE.h";

%template(vector_adouble) std::vector <adept::adouble>;
%template(vector_double) std::vector <double>;
```

I then make the changes following #2228 to change 
`typename internal::enable_if<E::is_active && E::rank==0, Active&>::type`
to 
`typename internal::enable_if<(E::is_active && E::rank==0), Active&>::type`

I am now able to parse through `Active.h`. However I error out at [`BinaryOperation.h`](https://github.com/rjhogan/Adept-2/blob/master/include/adept/BinaryOperation.h) with a syntax error

```
/usr/local/include/adept/Active.h:203: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:227: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:250: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:272: Warning 362: operator= ignored
/usr/local/include/adept/UnaryOperation.h:131: Warning 302: Identifier 'values_at_location_store_' redefined (ignored),
/usr/local/include/adept/UnaryOperation.h:124: Warning 302: previous definition of 'values_at_location_store_'.
/usr/local/include/adept/UnaryOperation.h:326: Warning 361: operator! ignored
/usr/local/include/adept/BinaryOperation.h:1116: Warning 302: Identifier 'operation' redefined (ignored),
/usr/local/include/adept/BinaryOperation.h:1110: Warning 302: previous definition of 'operation'.
/usr/local/include/adept/BinaryOperation.h:1196: Warning 302: Identifier 'operation' redefined (ignored),
/usr/local/include/adept/BinaryOperation.h:1191: Warning 302: previous definition of 'operation'.
/usr/local/include/adept/BinaryOperation.h:1372: Warning 302: Identifier '__truediv__' redefined (ignored) (Renamed from 'operator /'),
/usr/local/include/adept/BinaryOperation.h:1353: Warning 302: previous definition of '__truediv__' (Renamed from 'operator /').
/usr/local/include/adept/BinaryOperation.h:1449: Error: Syntax error in input(1).
make[2]: *** [CMakeFiles/simpleExample_swig_compilation.dir/build.make:76: CMakeFiles/_simpleExample.dir/simpleExamplePYTHON.stamp] Error 1
make[2]: *** Deleting file 'CMakeFiles/_simpleExample.dir/simpleExamplePYTHON.stamp'
make[1]: *** [CMakeFiles/Makefile2:111: CMakeFiles/simpleExample_swig_compilation.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

Line 1449 corresponds to :

`ADEPT_DEFINE_OPERATOR(GreaterThan, operator>, >, " > ")`

Is there a way to get rid of this syntax error?

**Comment by ojwb** [comment]
> Is there a way to get rid of this syntax error?

If you're able to change the header, wrapping `#ifndef SWIG` ... `#endif` around lines SWIG can't parse should hide them from the parser.  Obviously that's typically not possible if you're wrapping a third-party library.

If that's not suitable, it's hard to say without seeing what this expands to.  What does the preprocessed source for that line look like?

You don't seem to say what SWIG version you're using, but a newer SWIG version might also help as parsing improvements are fairly common.  For example SWIG 4.1.0 improved parsing of expressions containing `>` and `<` which could conceivably help here.  Building SWIG from git master might be worth trying (it's certainly what I'd try, but I do already have it checked out and built...)

**Comment by Huzaifg** [comment]
I am already using 4.1.1. which is the latest stable release version. I however tried using `#ifndef SWIG` ... `#endif` around all the lines that SWIG could not parse. I also expanded the interface file to import almost all adept dependencies

```
%module simpleExample

%include "std_string.i"
%include "std_vector.i"

%{
#include "adept.h"
#include "solveODE.h"
%}
%import "adept.h";
%import <adept/Active.h>
%import <adept/settings.h>
%import <adept/UnaryOperation.h>
%import <adept/BinaryOperation.h>
%import <adept/scalar_shortcuts.h>

%import <adept/base.h>
%import <adept/StackStorageOrig.h>
%import <adept/StackStorageOrigStl.h>
%import <adept/traits.h>
%import <adept/exception.h>
%import <adept/Statement.h>
%import <adept/Expression.h>
%import <adept/ArrayWrapper.h>
%import <adept/Stack.h>
%include "solveODE.h";

%template(vector_adouble) std::vector <adept::adouble>;
%template(vector_double) std::vector <double>;
```
I am now able to parse everything but I still end up with the same error as the first comment 



```
[ 25%] Swig compile simpleExample.i for python
/usr/local/include/adept/Active.h:203: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:227: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:250: Warning 362: operator= ignored
/usr/local/include/adept/Active.h:272: Warning 362: operator= ignored
/usr/local/include/adept/UnaryOperation.h:131: Warning 302: Identifier 'values_at_location_store_' redefined (ignored),
/usr/local/include/adept/UnaryOperation.h:124: Warning 302: previous definition of 'values_at_location_store_'.
/usr/local/include/adept/UnaryOperation.h:326: Warning 361: operator! ignored
/usr/local/include/adept/BinaryOperation.h:1116: Warning 302: Identifier 'operation' redefined (ignored),
/usr/local/include/adept/BinaryOperation.h:1110: Warning 302: previous definition of 'operation'.
/usr/local/include/adept/BinaryOperation.h:1196: Warning 302: Identifier 'operation' redefined (ignored),
/usr/local/include/adept/BinaryOperation.h:1191: Warning 302: previous definition of 'operation'.
/usr/local/include/adept/BinaryOperation.h:1372: Warning 302: Identifier '__truediv__' redefined (ignored) (Renamed from 'operator /'),
/usr/local/include/adept/BinaryOperation.h:1353: Warning 302: previous definition of '__truediv__' (Renamed from 'operator /').
/usr/local/include/adept/traits.h:127: Warning 325: Nested struct not currently supported (no ignored)
/usr/local/include/adept/Stack.h:74: Warning 389: operator[] ignored (consider using %extend)
/usr/local/include/adept/Stack.h:75: Warning 389: operator[] ignored (consider using %extend)
/usr/local/include/adept/Expression.h:24: Warning 315: Nothing known about 'internal::Packet'.
[ 25%] Built target simpleExample_swig_compilation
[ 50%] Building CXX object CMakeFiles/_simpleExample.dir/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx.o
In file included from /usr/local/include/adept.h:17,
                 from /home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:3664:
/usr/local/include/adept/Active.h: In instantiation of ‘adept::Active<T>::Active(const PType&, typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type*) [with PType = SwigValueWrapper<adept::Active<double> >; Type = double; typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type = void]’:
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:6921:15:   required from here
/usr/local/include/adept/Active.h:86:9: error: cannot convert ‘const SwigValueWrapper<adept::Active<double> >’ to ‘double’ in initialization
   86 |       : val_(rhs), gradient_index_(ADEPT_ACTIVE_STACK->register_gradient())
      |         ^~~~~~~~~
/usr/local/include/adept/Active.h: In instantiation of ‘adept::Active<T>::Active(const PType&, typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type*) [with PType = swig::SwigPySequence_Ref<adept::Active<double> >; Type = double; typename adept::internal::enable_if<adept::internal::is_not_expression<PType>::value>::type = void]’:
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:4769:30:   required from ‘void swig::assign(const SwigPySeq&, Seq*) [with SwigPySeq = SwigPySequence_Cont<adept::Active<double> >; Seq = std::vector<adept::Active<double> >]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:4791:12:   required from ‘static int swig::traits_asptr_stdseq<Seq, T>::asptr(PyObject*, sequence**) [with Seq = std::vector<adept::Active<double> >; T = adept::Active<double>; PyObject = _object; sequence = std::vector<adept::Active<double> >]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:4852:52:   required from ‘static int swig::traits_asptr<std::vector<T> >::asptr(PyObject*, std::vector<T>**) [with T = adept::Active<double>; PyObject = _object]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:3816:37:   required from ‘int swig::asptr(PyObject*, Type**) [with Type = std::vector<adept::Active<double> >; PyObject = _object]’
/home/unjhawala/temp/d4dof/interfaces/simpleExample/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx:6355:23:   required from here
/usr/local/include/adept/Active.h:86:9: error: cannot convert ‘const swig::SwigPySequence_Ref<adept::Active<double> >’ to ‘double’ in initialization
make[2]: *** [CMakeFiles/_simpleExample.dir/build.make:76: CMakeFiles/_simpleExample.dir/CMakeFiles/_simpleExample.dir/simpleExamplePYTHON_wrap.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/_simpleExample.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

**Comment by ojwb** [comment]
Sorry, I'm not sure I can help more as I don't know about `SwigPySequence_Ref`.

**Comment by ojwb** [comment]
> ```
> typename internal::enable_if<E::is_active && E::rank==0, Active&>::type
> ```
> 
> I think you're hitting #2228 there.

I've just pushed a fix for #2228 (82ee2f8cd4803d63af0e7ee79f834d28edf0abcc) so you don't won't need to workaround this once SWIG 4.2.0 is released.

**Comment by wsfulton** [comment]
Closing as this looks to be resolved. If not please re-open with details of what is wrong with latest SWIG.
